src/main/java/frc/robot/SwerveModule.java

```diff
@@ -0,0 +1,110 @@
+package frc.robot;
+
+import com.ctre.phoenix.sensors.CANCoder;
+import com.ctre.phoenix.sensors.SensorTimeBase;
+import com.revrobotics.RelativeEncoder;
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import frc.robot.Constants.ModuleConstants;
+import frc.robot.Constants.DriveConstants;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.RobotController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class SwerveModule {
+    private final CANSparkMax mDriveMotor;
+    private final CANSparkMax mSteerMotor;
+    
+    private final RelativeEncoder mDriveEncoder;
+    private final RelativeEncoder mSteerEncoder;
+
+    private final PIDController mSteerPidController;
+
+    private final CANCoder mAbsoluteEncoder;
+    private final boolean mAbsoluteEncoderReversed;
+    private final double mAbsoluteEncoderOffsetRad;
+
+    public SwerveModule(int driveMotorId, int steerMotorId, boolean driveMotorReversed, boolean steerMotorReversed,
+    int absoluteEncoderId, double absoluteEncoderOffset, boolean absoluteEncoderReversed) {
+    //absolute encoders
+    this.mAbsoluteEncoderOffsetRad = absoluteEncoderOffset;
+    this.mAbsoluteEncoderReversed = absoluteEncoderReversed;
+    mAbsoluteEncoder = new CANCoder(absoluteEncoderId); //default is degrees per second
+    mAbsoluteEncoder.configFeedbackCoefficient(ModuleConstants.kAbsoluteEncoderCountsPerMin2Rad, "rad", SensorTimeBase.PerSecond); //convert to radians per second
+    //motors
+    mDriveMotor = new CANSparkMax(driveMotorId, MotorType.kBrushless);
+    mSteerMotor = new CANSparkMax(steerMotorId, MotorType.kBrushless);
+    //inverting motors
+    mDriveMotor.setInverted(driveMotorReversed);
+    mSteerMotor.setInverted(steerMotorReversed);
+    //motor encoders
+    mDriveEncoder = mDriveMotor.getEncoder();
+    mSteerEncoder = mSteerMotor.getEncoder();
+
+    //so that we can work with meters and radians instead of rotations
+    mDriveEncoder.getPosition();
+    mDriveEncoder.setPositionConversionFactor(ModuleConstants.kDriveEncoderRotToMeters);
+    mDriveEncoder.setVelocityConversionFactor(ModuleConstants.kDriveEncoderRpm2Mps);
+    mSteerEncoder.setVelocityConversionFactor(ModuleConstants.kSteerEncoderRot2Rad);
+    mSteerEncoder.setVelocityConversionFactor(ModuleConstants.kSteerEncoderRPM2RadPerSec);
+    
+    //TODO: hardware PID?
+    
+    mSteerPidController = new PIDController(ModuleConstants.kPSteer, ModuleConstants.kISteer, ModuleConstants.kDSteer);
+    mSteerPidController.enableContinuousInput(-Math.PI, Math.PI);
+
+    
+    }
+
+    public double getDrivePosition() {
+        return mDriveEncoder.getPosition();
+    }
+
+    public double getSteerPosition() {
+        return mSteerEncoder.getPosition();
```

Comment:
You need to be offsetting this based on the `mAbsoluteEncoderOffsetRad` that you stored on construction. This would give you the raw position of the of steer sensor instead of the the direction of the steer wheel relative to forward.

What you'll find when getting this code working on the robot as that all the wheels when told to be point to `0.0` position, they will all point in different random directions. That's because when they installed the magnet and the encoder for the CANcoder, they have no way of making sure they are both oriented "forward". You'll have to find the offset needed to make sure that forward is `0.0`.

Reply:
getSteerPosition is for the relative encoder and not the cancoder. Should I be adding mAbsoluteEncoderOffsetRad to the cancoder or this relative encoder?

Reply:
So you are right, but you shouldn't be using the Spark Max's (conversely the Neo's) encoder. If you are doing a software PID, which you are in this version, the motor's encoder is almost entirely useless.